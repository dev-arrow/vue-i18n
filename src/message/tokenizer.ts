import {
  createScanner,
  CHAR_SP as SPACE,
  CHAR_LF as NEW_LINE
} from './scanner'
import {
  SourceLocation,
  createLocation,
  createPosition,
  Position
} from './location'

export const enum TokenTypes {
  Text, // 0
  Pipe,
  BraceLeft,
  BraceRight,
  Modulo,
  Named, // 5
  List,
  LinkedAlias,
  LinkedModifier,
  LinkedDelimiter,
  LinkedKey, // 10
  LinkedArg,
  ParenLeft,
  ParenRight,
  EOF
}

const enum TokenChars {
  Pipe = '|',
  BraceLeft = '{',
  BraceRight = '}',
  Modulo = '%',
  LinkedAlias = '@',
  LinkedModifier = '.',
  LinkedDelimiter = ':',
  ParenLeft = '(',
  ParenRight = ')'
}

// TODO: should be move to utils
const EOF = undefined

export type Token = {
  type: TokenTypes
  value?: string | number
  loc?: SourceLocation
}

export type Tokenizer = Readonly<{
  currentPosition: () => Position,
  nextToken: () => Token
}>

export function createTokenizer (source: string): Tokenizer {
  const _source = source
  const _scnr = createScanner(source)

  const currentPosition = (): Position => {
    return createPosition(_scnr.line(), _scnr.column(), _scnr.index())
  }

  let _currentType = TokenTypes.EOF
  let _currentValue: string | number | undefined | null = null
  let _startLoc = currentPosition()
  let _endLoc = _startLoc
  let _lastStartLoc: Position | null = null
  let _lastEndLoc: Position | null = null

  const getToken = (type: TokenTypes, value?: string | number): Token => {
    _endLoc = currentPosition()
    _currentType = type
    _currentValue = value
    return { type, value, loc: createLocation(_startLoc, _endLoc) }
  }

  const peekSpaces = (): void => {
    while (_scnr.currentPeek() === SPACE || _scnr.currentPeek() === NEW_LINE) {
      _scnr.peek()
    }
  }

  const skipSpaces = (): void => {
    peekSpaces()
    _scnr.skipToPeek()
  }

  const isIdentifierStart = (ch: string): boolean => {
    if (!ch) { return false }
    const cc = ch.charCodeAt(0)
    return (cc >= 97 && cc <= 122) || // a-z
           (cc >= 65 && cc <= 90) // A-Z
  }

  const isNumberStart = (ch: string): boolean => {
    if (!ch) {
      return false
    }
    const cc = ch.charCodeAt(0)
    return cc >= 48 && cc <= 57 // 0-9
  }

  const isNamedIdentifier = (): boolean => {
    if (_currentType !== TokenTypes.BraceLeft) { return false }
    peekSpaces()
    const ret = isIdentifierStart(_scnr.currentPeek())
    _scnr.resetPeek()
    return ret
  }

  const isListIdentifier = (): boolean => {
    if (_currentType !== TokenTypes.BraceLeft) { return false }
    peekSpaces()
    const ch = _scnr.currentPeek() === '-'
      ? _scnr.peek()
      : _scnr.currentPeek()
    const ret = isNumberStart(ch)
    _scnr.resetPeek()
    return ret
  }

  const isLinkedModifier = (): boolean => {
    if (_currentType !== TokenTypes.LinkedModifier) { return false }
    const ret = isIdentifierStart(_scnr.currentPeek())
    _scnr.resetPeek()
    return ret
  }

  const isLinkedIdentifier = (): boolean => {
    if (!(_currentType === TokenTypes.LinkedDelimiter ||
          _currentType === TokenTypes.ParenLeft)) {
      return false
    }
    const fn = (): boolean => {
      const ch = _scnr.currentPeek()
      if (ch === TokenChars.BraceLeft) {
        return isIdentifierStart(_scnr.peek())
      } else if ((ch === TokenChars.LinkedAlias ||
        ch === TokenChars.Modulo ||
        ch === TokenChars.Pipe ||
        ch === TokenChars.LinkedDelimiter ||
        ch === TokenChars.LinkedModifier ||
        ch === SPACE || !ch)) {
        return false
      } else if (ch === NEW_LINE) {
        return fn()
      } else { // other charactors
        return isIdentifierStart(ch)
      }
    }
    const ret = fn()
    _scnr.resetPeek()
    return ret
  }

  const isTextStart = (): boolean => {
    if (_currentType === TokenTypes.BraceLeft ||
        _currentType === TokenTypes.Pipe ||
        _currentType === TokenTypes.ParenLeft ||
        _currentType === TokenTypes.LinkedModifier ||
        _currentType === TokenTypes.LinkedDelimiter) {
      return false
    }

    const fn = (hasSpace = false): boolean => {
      const ch = _scnr.currentPeek()
      if ((ch === TokenChars.BraceLeft ||
        ch === TokenChars.Modulo ||
        ch === TokenChars.LinkedAlias || !ch)) {
        return hasSpace
      } else if (ch === TokenChars.Pipe) {
        return false
      } else if (ch === SPACE) {
        _scnr.peek()
        return fn(true)
      } else if (ch === NEW_LINE) {
        _scnr.peek()
        return fn(true)
      } else {
        return true
      }
    }

    const ret = fn()
    _scnr.resetPeek()
    return ret
  }

  const takeChar = (fn: Function): string | undefined | null => {
    const ch = _scnr.currentChar()
    if (ch === EOF) { return EOF }
    if (fn(ch)) {
      _scnr.next()
      return ch
    }
    return null
  }

  const takeIdentifierChar = (): string | undefined | null => {
    const closure = (ch: string) => {
      const cc = ch.charCodeAt(0)
      return ((cc >= 97 && cc <= 122) || // a-z
              (cc >= 65 && cc <= 90) || // A-Z
              (cc >= 48 && cc <= 57) || // 0-9
               cc === 95 || cc === 36) // _ $
    }
    return takeChar(closure)
  }

  const takeDigit = (): string | undefined | null => {
    const closure = (ch: string) => {
      const cc = ch.charCodeAt(0)
      return (cc >= 48 && cc <= 57) // 0-9
    }
    return takeChar(closure)
  }

  const getDigits = (): string => {
    let ch: string | undefined | null = ''
    let num = ''
    while ((ch = takeDigit())) {
      num += ch
    }
    if (num.length === 0) {
      // TODO: parse errror
    }
    return num
  }

  const readText = (): string => {
    const fn = (detect = false, buf: string): string => {
      const ch = _scnr.currentChar()
      if ((ch === TokenChars.BraceLeft ||
        ch === TokenChars.Modulo ||
        ch === TokenChars.LinkedAlias || !ch)) {
        return buf
      } else if (ch === TokenChars.Pipe) {
        return buf
      } else if (ch === SPACE) {
        buf += ch
        _scnr.next()
        return fn(detect, buf)
      } else if (ch === NEW_LINE) {
        buf += ch
        _scnr.next()
        return fn(detect, buf)
      } else {
        buf += ch
        _scnr.next()
        return fn(true, buf)
      }
    }
    return fn(false, '')
  }

  const readNamedIdentifier = (): string => {
    skipSpaces()
    let ch: string | undefined | null = ''
    let name = ''
    while ((ch = takeIdentifierChar())) {
      name += ch
    }
    skipSpaces()
    return name
  }

  const readListIdentifier = (): number => {
    skipSpaces()
    let value = ''
    if (_scnr.currentChar() === '-') {
      _scnr.next()
      value += `-${getDigits()}`
    } else {
      value += getDigits()
    }
    skipSpaces()
    return parseInt(value, 10)
  }

  const readLinkedModifierArg = (): string => {
    let ch: string | undefined | null = ''
    let name = ''
    while ((ch = takeIdentifierChar())) {
      name += ch
    }
    return name
  }

  const readLinkedIdentifier = (): string => {
    const fn = (detect = false, useParentLeft = false, buf: string): string => {
      const ch = _scnr.currentChar()
      if ((ch === TokenChars.BraceLeft ||
        ch === TokenChars.Modulo ||
        ch === TokenChars.LinkedAlias ||
        ch === TokenChars.ParenRight ||
        ch === TokenChars.Pipe || !ch)) {
        return buf
      } else if (ch === SPACE) {
        if (useParentLeft) {
          buf += ch
          _scnr.next()
          return fn(detect, useParentLeft, buf)
        } else {
          return buf
        }
      } else if (ch === NEW_LINE) {
        buf += ch
        _scnr.next()
        return fn(detect, useParentLeft, buf)
      } else {
        buf += ch
        _scnr.next()
        return fn(true, useParentLeft, buf)
      }
    }
    return fn(false, _currentType === TokenTypes.ParenLeft, '')
  }

  const readToken = (): Token => {
    let token = { type: TokenTypes.EOF }
    const ch = _scnr.currentChar()
    switch (ch) {
      case TokenChars.BraceLeft:
        _scnr.next()
        token = getToken(TokenTypes.BraceLeft, TokenChars.BraceLeft)
        break
      case TokenChars.BraceRight:
        _scnr.next()
        token = getToken(TokenTypes.BraceRight, TokenChars.BraceRight)
        break
      case TokenChars.LinkedAlias:
        _scnr.next()
        token = getToken(TokenTypes.LinkedAlias, TokenChars.LinkedAlias)
        break
      case TokenChars.LinkedModifier:
        _scnr.next()
        token = getToken(TokenTypes.LinkedModifier, TokenChars.LinkedModifier)
        break
      case TokenChars.LinkedDelimiter:
        _scnr.next()
        token = getToken(TokenTypes.LinkedDelimiter, TokenChars.LinkedDelimiter)
        break
      case TokenChars.ParenLeft:
        _scnr.next()
        token = getToken(TokenTypes.ParenLeft, TokenChars.ParenLeft)
        break
      case TokenChars.ParenRight:
        _scnr.next()
        token = getToken(TokenTypes.ParenRight, TokenChars.ParenRight)
        break
      case TokenChars.Modulo:
        _scnr.next()
        token = getToken(TokenTypes.Modulo, TokenChars.Modulo)
        break
      default:
        if (isTextStart()) {
          token = getToken(TokenTypes.Text, readText())
        } else if (isNamedIdentifier()) {
          token = getToken(TokenTypes.Named, readNamedIdentifier())
        } else if (isListIdentifier()) {
          token = getToken(TokenTypes.List, readListIdentifier())
        } else if (isLinkedModifier()) {
          token = getToken(TokenTypes.LinkedArg, readLinkedModifierArg())
        } else if (isLinkedIdentifier()) {
          if (ch === TokenChars.BraceLeft) {
            _scnr.next()
            token = getToken(TokenTypes.BraceLeft, TokenChars.BraceLeft)
          } else {
            token = getToken(TokenTypes.LinkedKey, readLinkedIdentifier())
          }
        }
        break
    }
    return token
  }

  const nextToken = (): Token => {
    _lastStartLoc = _startLoc
    _lastEndLoc = _endLoc
    _startLoc = currentPosition()

    if (!_scnr.currentChar()) {
      return getToken(TokenTypes.EOF)
    }

    return readToken()
  }

  return Object.freeze({
    nextToken, currentPosition
  })
}

export function parse (source: string): Token[] {
  const tokens = [] as Token[]
  const tokenizer = createTokenizer(source)
  let token: Token | null = null
  do {
    token = tokenizer.nextToken()
    tokens.push(token)
  } while (token.type !== TokenTypes.EOF)
  return tokens
}
