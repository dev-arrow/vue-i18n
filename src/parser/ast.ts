export const enum NodeTypes {
  RESOURCE,
  LITERAL,
  NAMED,
  LIST,
  PLURAL,
  LINKED,
  FUNCTION
}

// not containing whitespace or control characters
export type Identifier = string

export interface Node {
  type: NodeTypes
  loc: SourceLocation
}

// The node's range.
// The `start` is inclusive and `end` is exclusive.
// [start, end)
export interface SourceLocation {
  start: Position
  end: Position
  source: string
}

export interface Position {
  offset: number // from start of content (file)
  line: number
  column: number
}

type SinglePartNode =
  | LiteralNode
  | NamedNode
  | ListNode
  | LinkedNode
  | FunctionNode

type MultiPartNode = PluralNode

export interface ResourceNode extends Node {
  type: NodeTypes.RESOURCE
  body: SinglePartNode[] | MultiPartNode
}

export interface LiteralNode extends Node {
  type: NodeTypes.LITERAL
  value: string
}

export interface NamedNode extends Node {
  type: NodeTypes.NAMED
  key: Identifier
}

export interface ListNode extends Node {
  type: NodeTypes.LIST
  index: number
}

export interface PluralNode extends Node {
  type: NodeTypes.PLURAL
  cases: SinglePartNode[]
}

export interface LinkedNode extends Node {
  type: NodeTypes.LINKED
  key: Identifier | NamedNode | ListNode
}

export interface FunctionNode extends Node {
  type: NodeTypes.FUNCTION
  name: Identifier
  key: Identifier
  args?: Identifier[]
}
