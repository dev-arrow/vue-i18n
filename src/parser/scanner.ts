type Scanner = Readonly<{
  index: () => number
  peekOffset: () => number
  charAt: (offset: number) => string
  currentChar: () => string
  currentPeek: () => string
  next: () => string
  peek: () => string
  reset: () => void
  resetPeek: (offset?: number) => void
  skipToPeek: () => void
}>

export function createScanner (str: string): Scanner {
  const _buf = str
  let _index = 0
  let _peekOffset = 0

  const index = () => _index
  const peekOffset = () => _peekOffset

  function charAt (offset: number) {
    if (_buf[offset] === '\r' &&
        _buf[offset + 1] === '\n') {
      return '\n'
    }
    return _buf[offset]
  }

  const currentChar = () => charAt(_index)
  const currentPeek = () => charAt(_index + _peekOffset)

  function next () {
    _peekOffset = 0
    if (_buf[_index] === '\r' &&
        _buf[_index + 1] === '\n') {
      _index++
    }
    _index++
    return _buf[_index]
  }

  function peek () {
    if (_buf[_index + _peekOffset] === '\r' &&
        _buf[_index + _peekOffset + 1] === '\n') {
      _peekOffset++
    }
    _peekOffset++
    return _buf[_index + _peekOffset]
  }

  function reset () {
    _index = 0
    _peekOffset = 0
  }

  function resetPeek (offset = 0) {
    _peekOffset = offset
  }

  function skipToPeek () {
    _index += _peekOffset
    _peekOffset = 0
  }

  return Object.freeze({
    index, peekOffset, charAt, currentChar, currentPeek,
    next, peek, reset, resetPeek, skipToPeek
  })
}
