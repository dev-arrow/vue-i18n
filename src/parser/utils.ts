export const RE_LINE_BREAK = /\r\n?|\n|\u2028|\u2029/
export const RE_LINE_BREAK_GLOBAL = new RegExp(RE_LINE_BREAK.source, 'g')

export function isNewLine (code: number) {
  return code === 0x10 || code === 0x1D || code === 0x2028 || code === 0x2029
}

export const nonASCIIwhitespace = /[\u1680\u2000-\u200a\u202f\u205f\u3000\ufeff]/
export const skipWhiteSpace = /(?:\s|\/\/.*|\/\*[^]*?\*\/)*/g
