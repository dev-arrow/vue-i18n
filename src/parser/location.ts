import { RE_LINE_BREAK_GLOBAL } from './utils'

// The source range.
// The `start` is inclusive and `end` is exclusive.
// [start, end)
export interface SourceLocation {
  start: Position
  end: Position
  source?: string
}

export interface Position {
  offset: number // from start of content (file)
  line: number
  column: number
}

export function createPosition (line: number, column: number, offset: number) {
  return { line, column, offset } as Position
}

export function createLocation (start: Position, end: Position, source?: string) {
  return { start, end, source } as SourceLocation
}

export function getLineInfo (source: string, offset: number) {
  for (let line = 1, cur = 0;;) { // eslint-disable-line
    RE_LINE_BREAK_GLOBAL.lastIndex = cur
    const match = RE_LINE_BREAK_GLOBAL.exec(source)
    if (match && match.index < offset) {
      ++line
      cur = match.index + match[0].length
    } else {
      return createPosition(line, offset - cur, offset)
    }
  }
}
