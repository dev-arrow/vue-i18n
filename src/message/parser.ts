import { SourceLocation, Position } from './location'
import { createTokenizer, Tokenizer, TokenTypes, TokenizeContext } from './tokenizer'

export const enum NodeTypes {
  Resource, // 0
  Plural,
  Message,
  Text,
  Named,
  List, // 5
  Linked,
  LinkedKey,
  LinkedModifier
}

// not containing whitespace or control characters
export type Identifier = string

export interface Node {
  type: NodeTypes
  start: number
  end: number
  loc?: SourceLocation
}

export interface ResourceNode extends Node {
  type: NodeTypes.Resource
  body: MessageNode | PluralNode
}

export interface PluralNode extends Node {
  type: NodeTypes.Plural
  cases: MessageNode[]
}

export interface MessageNode extends Node {
  type: NodeTypes.Message
  items: MessageElementNode[]
}

type MessageElementNode =
  | TextNode
  | NamedNode
  | ListNode
  | LinkedNode

export interface TextNode extends Node {
  type: NodeTypes.Text
  value: string
}

export interface NamedNode extends Node {
  type: NodeTypes.Named
  key: Identifier
}

export interface ListNode extends Node {
  type: NodeTypes.List
  index: number
}

export interface LinkedNode extends Node {
  type: NodeTypes.Linked
  modifier?: LinkedModitierNode
  key: LinkedKeyNode | NamedNode | ListNode
}

export interface LinkedKeyNode extends Node {
  type: NodeTypes.LinkedKey
  value: string
}

export interface LinkedModitierNode extends Node {
  type: NodeTypes.LinkedModifier
  value: Identifier
}

type Parser = Readonly<{
  parse: (source: string) => ResourceNode
}>

export function createParser (): Parser {
  const startNode = (context: TokenizeContext, type: NodeTypes, init?: { offset: number, loc: Position }): Node => {
    let { lastOffset, lastStartLoc } = context
    if (init) {
      lastOffset = init.offset
      lastStartLoc = init.loc
    }
    return {
      type,
      start: lastOffset,
      end: -1,
      loc: { start: lastStartLoc, end: lastStartLoc }
    }
  }

  const endNode = (tokenizer: Tokenizer, node: Node, type?: NodeTypes): void => {
    node.end = tokenizer.currentOffset()
    if (type) {
      node.type = type
    }
    if (node.loc) {
      node.loc.end = tokenizer.currentPosition()
    }
  }

  const parseText = (tokenizer: Tokenizer, value: string): TextNode => {
    const context = tokenizer.context()
    const node = startNode(context, NodeTypes.Text) as TextNode
    node.value = value
    endNode(tokenizer, node)
    return node
  }

  const parseList = (tokenizer: Tokenizer, index: number): ListNode => {
    const context = tokenizer.context()
    const node = startNode(context, NodeTypes.List) as ListNode
    node.index = index
    // skip brach right
    tokenizer.nextToken()
    endNode(tokenizer, node)
    return node
  }

  const parseNamed = (tokenizer: Tokenizer, key: string): NamedNode => {
    const context = tokenizer.context()
    const node = startNode(context, NodeTypes.Named) as NamedNode
    node.key = key
    // skip brach right
    tokenizer.nextToken()
    endNode(tokenizer, node)
    return node
  }

  const parseLinkedModifier = (tokenizer: Tokenizer): LinkedModitierNode => {
    const token = tokenizer.nextToken()
    // asset check token
    if (!token.value || typeof token.value === 'number') {
      // TODO: should be thrown syntax error
      throw new Error()
    }
    const context = tokenizer.context()
    const node = startNode(context, NodeTypes.LinkedModifier) as LinkedModitierNode
    node.value = token.value
    endNode(tokenizer, node)
    return node
  }

  const parseLinkedKey = (tokenizer: Tokenizer, value: string): LinkedKeyNode => {
    const context = tokenizer.context()
    const node = startNode(context, NodeTypes.LinkedKey) as LinkedKeyNode
    node.value = value
    endNode(tokenizer, node)
    return node
  }

  const parseLinked = (tokenizer: Tokenizer): LinkedNode => {
    const context = tokenizer.context()
    const linkedNode = startNode(context, NodeTypes.Linked) as LinkedNode

    let token = tokenizer.nextToken()
    if (token.type === TokenTypes.LinkedDot) {
      linkedNode.modifier = parseLinkedModifier(tokenizer)
      token = tokenizer.nextToken()
    }

    // asset check token
    if (token.type !== TokenTypes.LinkedDelimiter) {
      // TODO: should be thrown syntax error
      throw new Error()
    }
    token = tokenizer.nextToken()

    // skip paren left
    if (token.type === TokenTypes.ParenLeft) {
      token = tokenizer.nextToken()
    }

    // skip brace left
    if (token.type === TokenTypes.BraceLeft) {
      token = tokenizer.nextToken()
    }

    switch (token.type) {
      case TokenTypes.LinkedKey:
        if (!token.value || typeof token.value !== 'string') {
          // TODO: should be thrown syntax error
          throw new Error()
        }
        linkedNode.key = parseLinkedKey(tokenizer, token.value)
        break
      case TokenTypes.Named:
        if (!token.value || typeof token.value === 'number') {
          // TODO: should be thrown syntax error
          throw new Error()
        }
        linkedNode.key = parseNamed(tokenizer, token.value)
        break
      case TokenTypes.List:
        if (token.value === undefined || typeof token.value === 'string') {
          // TODO: should be thrown syntax error
          throw new Error()
        }
        linkedNode.key = parseList(tokenizer, token.value)
        break
      default:
        // TODO: should be thrown syntax error
        throw new Error()
    }

    endNode(tokenizer, linkedNode)
    return linkedNode
  }

  const parseMessage = (tokenizer: Tokenizer): MessageNode => {
    const context = tokenizer.context()
    const node = startNode(context, NodeTypes.Message) as MessageNode
    node.items = []

    do {
      const token = tokenizer.nextToken()
      switch (token.type) {
        case TokenTypes.Text:
          if (!token.value || typeof token.value === 'number') {
            // TODO: should be thrown syntax error
            throw new Error()
          }
          node.items.push(parseText(tokenizer, token.value))
          break
        case TokenTypes.List:
          if (token.value === undefined || typeof token.value === 'string') {
            // TODO: should be thrown syntax error
            throw new Error()
          }
          node.items.push(parseList(tokenizer, token.value))
          break
        case TokenTypes.Named:
          if (!token.value || typeof token.value === 'number') {
            // TODO: should be thrown syntax error
            throw new Error()
          }
          node.items.push(parseNamed(tokenizer, token.value))
          break
        case TokenTypes.LinkedAlias:
          node.items.push(parseLinked(tokenizer))
          break
        default:
          break
      }
    } while (context.currentType !== TokenTypes.EOF && context.currentType !== TokenTypes.Pipe)

    endNode(tokenizer, node)
    return node
  }

  const parsePlural = (tokenizer: Tokenizer, offset: number, loc: Position, msgNode: MessageNode): PluralNode => {
    const context = tokenizer.context()

    const node = startNode(context, NodeTypes.Plural, { offset, loc }) as PluralNode
    node.cases = []
    node.cases.push(msgNode)

    do {
      const msg = parseMessage(tokenizer)
      node.cases.push(msg)
    } while (context.currentType !== TokenTypes.EOF)

    endNode(tokenizer, node)
    return node
  }

  const parseResource = (tokenizer: Tokenizer): MessageNode | PluralNode => {
    const context = tokenizer.context()
    const { offset, startLoc } = context

    const msgNode = parseMessage(tokenizer)
    if (context.currentType === TokenTypes.EOF) {
      return msgNode
    } else {
      return parsePlural(tokenizer, offset, startLoc, msgNode)
    }
  }

  function parse (source: string): ResourceNode {
    const tokenizer = createTokenizer(source)
    const context = tokenizer.context()

    const node = startNode(context, NodeTypes.Resource) as ResourceNode
    node.body = parseResource(tokenizer)

    // assert wheather achieved to EOF
    if (context.currentType !== TokenTypes.EOF) {
      // TODO: should be thrown syntax error
      throw new Error()
    }

    endNode(tokenizer, node)
    return node
  }

  return Object.freeze({
    parse
  })
}
