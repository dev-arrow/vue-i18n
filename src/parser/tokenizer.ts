import { createScanner, Scanner } from './scanner'
import { SourceLocation, createLocation, createPosition } from './location'

export const enum TokenTypes {
  Text, // 0
  Pipe,
  BraceL,
  BraceR,
  Modulo,
  Named, // 5
  List,
  Linked,
  EOF
}

const enum TokenChars {
  Pipe = '|',
  BraceL = '{',
  BraceR = '}',
  Modulo = '%',
  LinkedFirst = '@',
  LinkedLast = ':',
  EOL = '\n'
}

const EOF = undefined

export type Token = {
  type: TokenTypes
  value: string
  loc?: SourceLocation
}

type Tokenizer = Readonly<{
  parse: (source: string) => Token[]
}>

export function createTokenizer (): Tokenizer {
  function getBraceToken (scnr: Scanner, type: TokenTypes): Token {
    const start = createPosition(scnr.line(), scnr.column(), scnr.index())
    const value = scnr.next()
    return {
      type,
      value,
      loc: createLocation(
        start,
        createPosition(scnr.line(), scnr.column(), scnr.index())
      )
    }
  }

  function getTextToken (scnr: Scanner): Token | null {
    const start = createPosition(scnr.line(), scnr.column(), scnr.index())
    const createToken = (value: string): Token => ({
      type: TokenTypes.Text,
      value,
      loc: createLocation(
        start,
        createPosition(scnr.line(), scnr.column(), scnr.index())
      )
    })

    let buf = ''
    let ch = ''
    while ((ch = scnr.currentChar())) {
      if ((ch === TokenChars.BraceL ||
        ch === TokenChars.Pipe ||
        ch === TokenChars.LinkedFirst)) {
        if (buf.length > 0) {
          return createToken(buf)
        } else {
          return null
        }
      }

      if (ch === EOF) {
        return createToken(buf)
      }

      buf += ch
      scnr.next()
    }

    return createToken(buf)
  }

  function parse (source: string): Token[] {
    const tokens = [] as Token[]
    const scnr = createScanner(source)

    let ch: string | undefined = ''
    loop: while ((ch = scnr.currentChar())) { // eslint-disable-line no-labels
      let token: Token | null = null
      switch (ch) {
        case TokenChars.BraceL:
          // tokens.push(getBraceToken(scnr, TokenTypes.BraceL))
          break
        case TokenChars.BraceR:
          // tokens.push(getBraceToken(scnr, TokenTypes.BraceR))
          break
        case TokenChars.Pipe:
          break
        case TokenChars.LinkedFirst:
          break
        case TokenChars.Modulo:
          break
        case TokenChars.EOL:
          break
        default: // Text
          token = getTextToken(scnr)
          token && tokens.push(token)
          break
      }
    }

    tokens.push({
      type: TokenTypes.EOF,
      value: '',
      loc: createLocation(
        createPosition(scnr.line(), scnr.column(), scnr.index()),
        createPosition(scnr.line(), scnr.column(), scnr.index())
      )
    })

    return tokens
  }

  return Object.freeze({
    parse
  })
}
