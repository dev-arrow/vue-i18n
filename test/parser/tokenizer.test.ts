import { createTokenizer } from '../../src/parser/tokenizer'

test('Text only', () => {
  const tokenizer = createTokenizer()

  const tokens1 = tokenizer.parse('hello world!')
  expect(tokens1).toMatchSnapshot('basic')

  const tokens2 = tokenizer.parse('hello\nworld!')
  expect(tokens2).toMatchSnapshot('multiline')
})

/*
test('plural', () => {
  const tokenizer = createTokenizer()

  const tokens1 = tokenizer.parse('no apples | one apple  |  too much apples  ')
  expect(tokens1).toMatchSnapshot('basic')

  const tokens2 = tokenizer.parse('no apples |\n one apple |\n too much apples')
  expect(tokens2).toMatchSnapshot('multiline')
})
*/

/*
test('named placeholder', () => {
  const tokenizer = createTokenizer()

  const tokens1 = tokenizer.parse('hi {name}!')
  expect(tokens1).toMatchSnapshot('basic')

  const tokens2 = tokenizer.parse('{last} {middle} {first}')
  expect(tokens2).toMatchSnapshot('multiple')

  const tokens3 = tokenizer.parse('hi {  name }!')
  expect(tokens3).toMatchSnapshot('spaces')

  const tokens4 = tokenizer.parse(`lastName: {last}\nmiddleName: {middle}\r\nfirstName: {first}`)
  expect(tokens4).toMatchSnapshot('multline')
})
*/
