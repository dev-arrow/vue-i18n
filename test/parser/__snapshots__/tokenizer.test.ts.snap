// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text only: basic 1`] = `
Array [
  Object {
    "loc": Object {
      "end": Object {
        "column": 13,
        "line": 1,
        "offset": 12,
      },
      "source": undefined,
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "type": 0,
    "value": "hello world!",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 13,
        "line": 1,
        "offset": 12,
      },
      "source": undefined,
      "start": Object {
        "column": 13,
        "line": 1,
        "offset": 12,
      },
    },
    "type": 8,
    "value": "",
  },
]
`;

exports[`Text only: multiline 1`] = `
Array [
  Object {
    "loc": Object {
      "end": Object {
        "column": 7,
        "line": 2,
        "offset": 12,
      },
      "source": undefined,
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "type": 0,
    "value": "hello
world!",
  },
  Object {
    "loc": Object {
      "end": Object {
        "column": 7,
        "line": 2,
        "offset": 12,
      },
      "source": undefined,
      "start": Object {
        "column": 7,
        "line": 2,
        "offset": 12,
      },
    },
    "type": 8,
    "value": "",
  },
]
`;
